src/main/scala/gemmini/CounterFile.scala

```diff
@@ -0,0 +1,171 @@
+  val ACC_D_WAIT_CYCLE = 30.U
+
+  val A_GARBAGE_CYCLE = 31.U
+  val B_GARBAGE_CYCLE = 32.U
+  val C_GARBAGE_CYCLE = 33.U
+
+  val IM2COL_MEM_CYCLES = 34.U
+  val IM2COL_ACTIVE_CYCLES = 35.U
+  val IM2COL_TRANSPOSER_WAIT_CYCLE = 36.U
+
+  val ROB_FULL_CYCLES = 37.U
+  val ROB_ACTIVE_CYCLES = 38.U
+
+  val LOOP_MATMUL_ACTIVE_CYCLES = 39.U
+  val TRANSPOSE_PRELOAD_UNROLLER_ACTIVE_CYCLES = 40.U
+
+  val n = 41
+}
+
+object CounterExternal {
+  val DISABLE = 0.U
+
+  val ROB_LD_COUNT = 1.U
+  val ROB_ST_COUNT = 2.U
+  val ROB_EX_COUNT = 3.U
+
+  val RDMA_BYTE_REC = 4.U
+  val WDMA_BYTE_SENT = 5.U
+
+  val n = 6
+}
+
+class CounterEventIO extends Bundle {
+  val event_signal = Input(Vec(CounterEvent.n, Bool()))
+  val external_values = Input(Vec(CounterExternal.n, p(XLen)))
+  val external_reset = Output(Bool())
+
+  // Connect Event Signal
+  private var connected = Array.fill(CounterEvent.n)(false)
+  def connectEventSignal(addr: UInt, sig: UInt) = {
+    event_signal(addr) := sig
+    connected(addr) = true
+  }
+
+  // Connect Event Signal
+  private var connected_external = Array.fill(CounterEvent.n)(false)
+  def connectExternalCounter(addr: UInt, ext_value: UInt) = {
+    external_values(addr) := ext_value
+    connected_external(addr) = true
+  }
+
+  // Collect IO from submodule
+  def collect(io: CounterIO) = {
+    io.external_reset := external_reset
+    for (i <- 0 to CounterEvent.n)
+      if (io.connected(i)) {
+        if (connected(i))
+          throw new IllegalStateException("Port " + i + " is already connected in another IO")
+        else {
+          connected(i) = true
+          event_signal(i) := io.event_signal(i)
+        }
+      }
+    for (i <- 0 to CounterExternal.n)
+      if (io.connected_external(i)) {
+        if (connected_external(i))
+          throw new IllegalStateException("External counter " + i + " is already connected in another IO")
+        else {
+          connected_external(i) = true
+          external_values(i) := io.external_values(i)
+        }
+      }
+  }
+}
+
+class CounterIO(nPerfCounter: Int, counterWidth: Int) extends Bundle {
+  val addrWidth = log2Ceil(nPerfCounter)
+
+  val reset = Input(Bool())
+  val snapshot = Input(Bool())
+  val addr = Input(UInt(addrWidth.W))
+  val data = Output(UInt(counterWidth.W))
+  val event_io = new CounterEventIO
+  val config_address = Valid(UInt(log2Ceil(CounterEvent.n)))
+}
+
+// A simple counter file. Every counter is incremented when the corresponding event signal is high on rising edge.
+class CounterFile(nPerfCounter: Int, counterWidth: Int) extends Module
+{
+  val io = IO(new CounterIO(nPerfCounter, counterWidth))
+
+  val counters = Vec(nPerfCounter, RegInit(0.U(counterWidth.W), io.reset))
+  val counter_config = Vec(nPerfCounter, RegInit(0.U(CounterEvent.n)))
+  val snapshot_enable = RegInit(false.B, io.reset)
+
+  // Snapshot: In case the access instruction get blocked, it is possible to take a snapshot when reading counter
```

Comment:
Why would the access instruction be "blocked"?

Reply:
This should be "a sequence of access instructions", I will fix the comment. If we are simulating with an OS, the thread can be preempted before we finish the sequence, making correlating the values of different counter difficult.

Reply:
Ah, I see, that's good thinking